File: base/src/main/java/com/boilerplate/base/mvp/BaseFragment.java
package com.boilerplate.base.mvp;

import android.app.ProgressDialog;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

public abstract class BaseFragment<T extends IPresenter> extends DialogFragment implements IView {


    protected ProgressDialog progressDialog;
    protected T mPresenter;
    protected View mView;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        mView = inflater.inflate(getLayoutId(), null);
        setupActivityComponent();
        mPresenter = createPresenter();
        return mView;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mPresenter.attachView(this);
        initEventAndData();
    }


    protected abstract int getLayoutId();

    protected abstract void initEventAndData();

    protected abstract T createPresenter();

    protected abstract void setupActivityComponent();
}

File: base/src/main/java/com/boilerplate/base/kotlin/BaseFragment.kt
package com.boilerplate.base.kotlin

import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.boilerplate.base.mvp.IPresenter
import com.boilerplate.base.mvp.IView

/**
 * Created by Louis on 2018/4/30.
 */
abstract class BaseFragment<V : IView, T : IPresenter<V>> : Fragment(), IView {
    protected abstract var mPresenter: T
    protected lateinit var mView: View
    protected abstract var layoutId: Int

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        mView = inflater.inflate(layoutId, null)
        setupActivityComponent()
        return mView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mPresenter.attachView(this as V)


    }

    protected abstract fun setupActivityComponent()

    protected abstract fun initEventAndData()

}